src/interpreter.rs

```diff
@@ -120,3 +129,1207 @@ bitflags::bitflags! {
         const CHECKLOCKTIMEVERIFY = 1 << 9;
     }
 }
+
+pub trait SignatureChecker {
+    fn check_sig(&self, _script_sig: &[u8], _vch_pub_key: &[u8], _script_code: &Script) -> bool {
+        false
+    }
+
+    fn check_lock_time(&self, _lock_time: &ScriptNum) -> bool {
+        false
+    }
+}
+
+pub struct BaseSignatureChecker();
+
+impl SignatureChecker for BaseSignatureChecker {}
+
+pub struct CallbackTransactionSignatureChecker<'a> {
+    pub sighash: SighashCalculator<'a>,
+    pub lock_time: &'a ScriptNum,
+    pub is_final: bool,
+}
+
+type ValType = Vec<u8>;
+
+fn set_success<T>(res: T) -> Result<T, ScriptError> {
+    Ok(res)
+}
+
+fn set_error<T>(serror: ScriptError) -> Result<T, ScriptError> {
+    Err(serror)
+}
+
+fn cast_to_bool(vch: &ValType) -> bool {
+    for i in 0..vch.len() {
+        if vch[i] != 0 {
+            // Can be negative zero
+            if i == vch.len() - 1 && vch[i] == 0x80 {
+                return false;
+            }
+            return true;
+        }
+    }
+    false
+}
+
+/**
+ * Script is a stack machine (like Forth) that evaluates a predicate
+ * returning a bool indicating valid or not.  There are no loops.
+ */
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Stack<T>(Vec<T>);
+
+/// Wraps a Vec (or whatever underlying implementation we choose in a way that matches the C++ impl
+/// and provides us some decent chaining)
+impl<T> Stack<T> {
+    fn reverse_index(&self, i: isize) -> Result<usize, ScriptError> {
+        usize::try_from(-i)
+            .map(|a| self.0.len() - a)
+            .map_err(|_| ScriptError::InvalidStackOperation)
```

Comment:
This allows `reverse_index(0)` which should probably be an error. It doesn't matter much, because `top(0)` will be an error.